### DISCUSSION TITLE
Reduce_ex error on add_punctuation_capitalization

### DISCUSSION THREAD

**Discussion by iry47** [NVIDIA-NeMo/NeMo#2424] [Q&A]
Hi, I wrote a jupyter notebook that loads my pretrained punctuation model from googe coloab and applies it in my application. However when I actually transfer the code into the application, I get stuck on an error :

File "/Users/ryanheadley/Sites/pomelo/backend/pomelo_backend/video.py", line 749, in run_vosk
    punctuatedSentences = punctuation_model.add_punctuation_capitalization(allText)
  File "/Users/ryanheadley/Sites/pomelo/backend/nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py", line 403, in add_punctuation_capitalization
    for batch in infer_datalayer:
  File "/Users/ryanheadley/.local/share/virtualenvs/backend-2Yznqhll/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 359, in __iter__
    return self._get_iterator()
  File "/Users/ryanheadley/.local/share/virtualenvs/backend-2Yznqhll/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 305, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "/Users/ryanheadley/.local/share/virtualenvs/backend-2Yznqhll/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 918, in __init__
    w.start()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
NotImplementedError: object proxy must define __reduce_ex__()

Any ideas on how to solve this?

**Comment by iry47** [comment]
Nobody has any ideas? I can't find a way to get around it and it prevents me from using the model... It works fine in my jupyter notebook but breaks in my Flask application @titu1994 @ekmb

**Reply by okuchaiev** [reply]
it looks like the code works in colab but not in your application? If this is correct, there could be some incompatibilities with your application environment.  Can you share the details about your application env?

**Reply by iry47** [reply]
Hi @okuchaiev, thanks for the reply. 
I'm working in a python 3.9.5 Flask application running in a pipenv. Here's my package list 

absl-py                       0.13.0
aiohttp                       3.7.4.post0
alabaster                     0.7.12
aniso8601                     9.0.1
antlr4-python3-runtime        4.8
appdirs                       1.4.4
asttokens                     2.0.5
async-timeout                 3.0.1
attrs                         21.2.0
audioread                     2.1.9
Babel                         2.9.0
bcrypt                        3.2.0
blinker                       1.4
boto3                         1.17.102
botocore                      1.20.102
braceexpand                   0.1.7
cachetools                    4.2.2
Cerberus                      1.3.3
certifi                       2020.12.5
cffi                          1.14.5
chardet                       4.0.0
cld2-cffi                     0.1.4
click                         7.1.2
colorama                      0.4.4
coverage                      5.5
cryptography                  3.4.7
cycler                        0.10.0
Cython                        0.29.23
decorator                     5.0.9
Distance                      0.1.3
docopt                        0.6.2
docutils                      0.16
editdistance                  0.5.3
executing                     0.6.0
ffmpeg-python                 0.2.0
filelock                      3.0.12
Flask                         1.1.2
Flask-Cors                    3.0.10
Flask-JWT-Extended            3.25.0
Flask-Limiter                 1.4
Flask-Mail                    0.9.1
Flask-RESTful                 0.3.8
Flask-SQLAlchemy              2.5.1
Flask-Testing                 0.8.1
frozendict                    2.0.3
fsspec                        2021.6.1
future                        0.18.2
fuzzywuzzy                    0.18.0
g2p-en                        2.1.0
gdown                         3.13.0
gensim                        4.0.1
google-api-core               1.26.3
google-auth                   1.30.0
google-auth-oauthlib          0.4.4
google-cloud-core             1.6.0
google-cloud-speech           2.3.0
google-cloud-storage          1.38.0
google-crc32c                 1.1.2
google-resumable-media        1.2.0
googleapis-common-protos      1.53.0
greenlet                      1.0.0
grpcio                        1.37.0
gspread                       3.7.0
gunicorn                      20.1.0
h5py                          3.3.0
httplib2                      0.19.1
huggingface-hub               0.0.12
hydra-core                    1.1.0
icecream                      2.1.0
idna                          2.10
imagesize                     1.2.0
inflect                       5.3.0
iniconfig                     1.1.1
ipadic                        1.0.0
itsdangerous                  1.1.0
jieba                         0.42.1
Jinja2                        2.11.3
jmespath                      0.10.0
joblib                        1.0.1
kaldiio                       2.17.2
kiwisolver                    1.3.1
libcst                        0.3.18
librosa                       0.8.1
limits                        1.5.1
llvmlite                      0.36.0
Markdown                      3.3.4
MarkupSafe                    1.1.1
matplotlib                    3.4.2
mecab-python3                 1.0.3
megatron-lm                   2.2.0
mock-alchemy                  0.2.2
mpmath                        1.2.1
multi-rake                    0.0.1
multidict                     5.1.0
mypy-extensions               0.4.3
mysql-connector-python        8.0.24
nemo-toolkit                  1.0.2
nltk                          3.6.2
nose2                         0.10.0
numba                         0.53.1
numpy                         1.21.0
numpy-indexed                 0.3.5
oauth2client                  4.1.3
oauthlib                      3.1.0
omegaconf                     2.1.0
onnx                          1.9.0
OpenCC                        1.1.1
packaging                     20.9
pandas                        1.2.4
pangu                         4.0.6.1
pickle5                       0.0.11
Pillow                        8.2.0
pip                           21.0.1
pluggy                        0.13.1
pooch                         1.4.0
portalocker                   2.0.0
proto-plus                    1.18.1
protobuf                      3.15.8
py                            1.10.0
pyannote.core                 4.1
pyannote.database             4.1
pyannote.metrics              3.0.1
pyasn1                        0.4.8
pyasn1-modules                0.2.8
pybind11                      2.6.2
pycparser                     2.20
pyDeprecate                   0.3.0
pydub                         0.25.1
pyemd                         0.5.1
Pygments                      2.8.1
PyJWT                         1.7.1
PyMySQL                       1.0.2
pyparsing                     2.4.7
pyrsistent                    0.17.3
PySocks                       1.7.1
pytest                        6.2.4
python-dateutil               2.8.1
python-Levenshtein            0.12.2
pytorch-lightning             1.3.7.post0
pytz                          2021.1
PyYAML                        5.4.1
rapidfuzz                     1.4.1
regex                         2021.4.4
requests                      2.25.1
requests-oauthlib             1.3.0
resampy                       0.2.2
rsa                           4.7.2
ruamel.yaml                   0.17.10
ruamel.yaml.clib              0.2.4
s3transfer                    0.4.2
sacrebleu                     1.5.1
sacremoses                    0.0.45
scikit-learn                  0.24.2
scipy                         1.6.3
sentencepiece                 0.1.96
sentry-sdk                    1.0.0
setuptools                    57.0.0
shellingham                   1.4.0
simplejson                    3.17.2
six                           1.15.0
smart-open                    5.0.0
snowballstemmer               2.1.0
sortedcontainers              2.4.0
SoundFile                     0.10.3.post1
Sphinx                        3.5.4
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        1.0.3
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.4
SQLAlchemy                    1.4.11
srt                           3.4.1
sympy                         1.8
tabulate                      0.8.9
tensorboard                   2.4.1
tensorboard-plugin-wit        1.8.0
textdistance                  4.2.1
threadpoolctl                 2.1.0
tokenizers                    0.10.3
toml                          0.10.2
torch                         1.9.0
torch-stft                    0.1.4
torchmetrics                  0.3.2
tqdm                          4.60.0
transformers                  4.8.1
typer                         0.3.2
typing-extensions             3.7.4.3
typing-inspect                0.6.0
Unidecode                     1.2.0
urllib3                       1.26.4
vosk                          0.3.21
webdataset                    0.1.62
websockets                    9.1
Werkzeug                      1.0.1
wget                          3.2
wheel                         0.36.2
wrapt                         1.12.1
yarl                          1.6.3
youtokentome                  1.0.6

Let me know if there's any other information that can help
